src/mlpack/methods/ann/gan_impl.hpp

```diff
@@ -0,0 +1,283 @@
+    reset(false)
+{
+  // Insert IdentityLayer for joining the Generator and Discriminator.
+  discriminator.network.insert(
+      discriminator.network.begin(),
+      new IdentityLayer<>());
+
+  counter = 0;
+  currentBatch = 0;
+
+  discriminator.deterministic = generator.deterministic = true;
+
+  responses.set_size(1, predictors.n_cols);
+  responses.ones();
+
+  discriminator.predictors.set_size(predictors.n_rows, predictors.n_cols + 1);
+  discriminator.predictors.cols(0, predictors.n_cols - 1) = predictors;
+
+  discriminator.responses.set_size(1, predictors.n_cols + 1);
+  discriminator.responses.ones();
+  discriminator.responses(predictors.n_cols) = 0;
+
+  numFunctions = predictors.n_cols;
+
+  noise.set_size(noiseDim, 1);
+
+  generator.predictors.set_size(noiseDim, noise.n_cols + 1);
+  generator.responses.set_size(predictors.n_rows, noise.n_cols + 1);
+}
+
+template<typename Model, typename InitializationRuleType, typename Noise>
+void GAN<Model, InitializationRuleType, Noise>::Reset()
+{
+  size_t genWeights = 0;
+  size_t discWeights = 0;
+
+  NetworkInitialization<InitializationRuleType> networkInit(initializeRule);
+
+  for (size_t i = 0; i < generator.network.size(); ++i){
+    genWeights += boost::apply_visitor(weightSizeVisitor, generator.network[i]);
+  }
+
+  for (size_t i = 0; i < discriminator.network.size(); ++i){
+    discWeights += boost::apply_visitor(weightSizeVisitor,
+        discriminator.network[i]);
+  }
+
+  parameter.set_size(genWeights + discWeights, 1);
+  generator.Parameters() = arma::mat(parameter.memptr(), genWeights, 1, false,
+      false);
+  discriminator.Parameters() = arma::mat(parameter.memptr() + genWeights,
+      discWeights, 1 , false, false);
+
+  // Initialize the parameters generator
+  networkInit.Initialize(generator.network, parameter);
+  // Initialize the parameters discriminator
+  networkInit.Initialize(discriminator.network, parameter, genWeights);
+
+  reset = true;
+}
+
+template<typename Model, typename InitializationRuleType, typename Noise>
+template<typename OptimizerType>
+void GAN<Model, InitializationRuleType, Noise>::Train(
+    OptimizerType& Optimizer)
+{
+  if (!reset)
+    Reset();
+  Optimizer.Optimize(*this, parameter);
+}
+
+template<typename Model, typename InitializationRuleType, typename Noise>
+double GAN<Model, InitializationRuleType, Noise>::Evaluate(
+    const arma::mat& /*parameters*/,
+    const size_t i,
+    const size_t /*batchSize*/)
+{
+  if (!reset)
+    Reset();
+
+  currentInput = this->predictors.unsafe_col(i);
+  currentTarget = this->responses.unsafe_col(i);
+
+  discriminator.Forward(std::move(currentInput));
+  double res = discriminator.outputLayer.Forward(
+      std::move(boost::apply_visitor(
+          outputParameterVisitor,
+          discriminator.network.back())), std::move(currentTarget));
+
+  noise.imbue( [&]() { return noiseFunction();} );
+  generator.Forward(std::move(noise));
+  arma::mat temp = boost::apply_visitor(
```

Comment:
What about using ``discriminator.predictors.col(numFunctions) = ...`` here instead of an extra temp parameter.

Reply:
Sure, that would be an optimization.